File: ontrack-service/src/main/java/net/nemerosa/ontrack/service/ValidationRunStatusServiceImpl.java
package net.nemerosa.ontrack.service;

import net.nemerosa.ontrack.model.exceptions.ValidationRunStatusChangeForbiddenException;
import net.nemerosa.ontrack.model.exceptions.ValidationRunStatusNotFoundException;
import net.nemerosa.ontrack.model.exceptions.ValidationRunStatusUnknownDependencyException;
import net.nemerosa.ontrack.model.structure.ValidationRunStatusID;
import net.nemerosa.ontrack.model.structure.ValidationRunStatusService;
import net.nemerosa.ontrack.model.support.StartupService;
import org.apache.commons.lang3.StringUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;

import java.util.*;
import java.util.stream.Collectors;

import static net.nemerosa.ontrack.model.structure.ValidationRunStatusID.*;

@Service
public class ValidationRunStatusServiceImpl implements ValidationRunStatusService, StartupService {

    private final Logger logger = LoggerFactory.getLogger(ValidationRunStatusService.class);

    private final Map<String, ValidationRunStatusID> statuses = new LinkedHashMap<>();

    @Override
    public Collection<ValidationRunStatusID> getValidationRunStatusList() {
        return statuses.values();
    }

    @Override
    public ValidationRunStatusID getValidationRunStatus(String id) {
        return Optional.ofNullable(statuses.get(id)).orElseThrow(() -> new ValidationRunStatusNotFoundException(id));
    }

    @Override
    public List<ValidationRunStatusID> getNextValidationRunStatusList(String id) {
        return getValidationRunStatus(id).getFollowingStatuses().stream()
                .map(this::getValidationRunStatus)
                .collect(Collectors.toList());
    }

    @Override
    public void checkTransition(ValidationRunStatusID from, ValidationRunStatusID to) {
        if (!from.getFollowingStatuses().contains(to.getId())) {
            throw new ValidationRunStatusChangeForbiddenException(from.getId(), to.getId());
        }
    }

    @Override
    public String getName() {
        return "Loading of validation run statuses";
    }

    @Override
    public int startupOrder() {
        return SYSTEM_REGISTRATION;
    }

    /**
     * Registers the tree of validation run status ids.
     */
    @Override
    public void start() {
        register(STATUS_PASSED);
        register(STATUS_FIXED);
        register(STATUS_DEFECTIVE);
        register(STATUS_EXPLAINED, FIXED);
        register(STATUS_INVESTIGATING, DEFECTIVE, EXPLAINED, FIXED);
        register(STATUS_INTERRUPTED, INVESTIGATING, FIXED);
        register(STATUS_FAILED, INTERRUPTED, INVESTIGATING, EXPLAINED, DEFECTIVE);
        register(STATUS_WARNING, INTERRUPTED, INVESTIGATING, EXPLAINED, DEFECTIVE);
        // TODO Participation from extensions
        // Checks the tree
        for (ValidationRunStatusID statusID : statuses.values()) {
            for (String nextStatus : statusID.getFollowingStatuses()) {
                if (!statuses.containsKey(nextStatus)) {
                    throw new ValidationRunStatusUnknownDependencyException(statusID.getId(), nextStatus);
                }
            }
        }
        // Logging
        for (ValidationRunStatusID statusID : statuses.values()) {
            logger.info(
                    "[status] {} -> {}",
                    statusID.getId(),
                    StringUtils.join(statusID.getFollowingStatuses(), ",")
            );
        }
    }

    private void register(ValidationRunStatusID statusID, String... next) {
        statuses.put(statusID.getId(), statusID.addDependencies(next));
    }

}

File: ontrack-service/src/main/java/net/nemerosa/ontrack/service/ValidationRunStatusServiceImpl.kt
package net.nemerosa.ontrack.service

import net.nemerosa.ontrack.model.exceptions.ValidationRunStatusChangeForbiddenException
import net.nemerosa.ontrack.model.exceptions.ValidationRunStatusNotFoundException
import net.nemerosa.ontrack.model.exceptions.ValidationRunStatusUnknownDependencyException
import net.nemerosa.ontrack.model.structure.ValidationRunStatusID
import net.nemerosa.ontrack.model.structure.ValidationRunStatusService
import net.nemerosa.ontrack.model.support.StartupService
import org.apache.commons.lang3.StringUtils
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import java.util.*

@Service
class ValidationRunStatusServiceImpl : ValidationRunStatusService, StartupService {

    private val logger = LoggerFactory.getLogger(ValidationRunStatusService::class.java)

    private val statuses: MutableMap<String, ValidationRunStatusID> = LinkedHashMap()

    override fun getValidationRunStatusList(): Collection<ValidationRunStatusID> = statuses.values

    override fun getValidationRunStatus(id: String): ValidationRunStatusID =
            Optional.ofNullable(statuses[id]).orElseThrow { ValidationRunStatusNotFoundException(id) }

    override fun getNextValidationRunStatusList(id: String): List<ValidationRunStatusID> =
            getValidationRunStatus(id)
                    .followingStatuses
                    .map { getValidationRunStatus(it) }

    override fun checkTransition(from: ValidationRunStatusID, to: ValidationRunStatusID) {
        if (!from.followingStatuses.contains(to.id)) {
            throw ValidationRunStatusChangeForbiddenException(from.id, to.id)
        }
    }

    override fun getName(): String = "Loading of validation run statuses"

    override fun startupOrder(): Int = StartupService.SYSTEM_REGISTRATION

    /**
     * Registers the tree of validation run status ids.
     */
    override fun start() {
        register(ValidationRunStatusID.STATUS_PASSED)
        register(ValidationRunStatusID.STATUS_FIXED)
        register(ValidationRunStatusID.STATUS_DEFECTIVE)
        register(ValidationRunStatusID.STATUS_EXPLAINED, ValidationRunStatusID.FIXED)
        register(ValidationRunStatusID.STATUS_INVESTIGATING, ValidationRunStatusID.DEFECTIVE, ValidationRunStatusID.EXPLAINED, ValidationRunStatusID.FIXED)
        register(ValidationRunStatusID.STATUS_INTERRUPTED, ValidationRunStatusID.INVESTIGATING, ValidationRunStatusID.FIXED)
        register(ValidationRunStatusID.STATUS_FAILED, ValidationRunStatusID.INTERRUPTED, ValidationRunStatusID.INVESTIGATING, ValidationRunStatusID.EXPLAINED, ValidationRunStatusID.DEFECTIVE)
        register(ValidationRunStatusID.STATUS_WARNING, ValidationRunStatusID.INTERRUPTED, ValidationRunStatusID.INVESTIGATING, ValidationRunStatusID.EXPLAINED, ValidationRunStatusID.DEFECTIVE)
        // Checks the tree
        for (statusID in statuses.values) {
            for (nextStatus in statusID.followingStatuses) {
                if (!statuses.containsKey(nextStatus)) {
                    throw ValidationRunStatusUnknownDependencyException(statusID.id, nextStatus)
                }
            }
        }
        // Logging
        for (statusID in statuses.values) {
            logger.info(
                    "[status] {} -> {}",
                    statusID.id,
                    StringUtils.join(statusID.followingStatuses, ",")
            )
        }
    }

    private fun register(statusID: ValidationRunStatusID, vararg next: String) {
        statuses[statusID.id] = statusID.addDependencies(*next)
    }
}